spacemeshos/smapp
# Implement basic and simplest auto-updating mechanism

It closes #863.

<img width="1392" alt="Screenshot 2022-02-12 at 02 05 39" src="https://user-images.githubusercontent.com/1897530/153677578-3ca8ac49-8cd6-4f60-b1e6-a379a88a0adc.png">

### How Smapp checks update
- on unlocking the wallet, switching the network, or each 24 hours
- it checks the [discovery service](https://discover.spacemesh.io/networks.json)
- compare `latestSmappRelease` (of current network) with the current version of Smapp
- if the current version differs — check for existing of `latest.yml` and artifact on `${smappBaseDownloadUrl}/${latestSmappRelease}`. E.G. `http://localhost:8000/smapp/0.2.2`

Once the update is downloaded Smapp will install the update on the first Smapp restart.

### How to test:
1. I built two signed releases (see release drafts) that can be updated one to another: `0.2.1-update.0` and `0.2.2`
2. Download the first one and install it as usual.
3. You need a patched networks.json (at least until we update it on discovery service). For testing purposes, I've added a possibility to set a custom URL to networks.json via argument `--discovery`. So run Spacemesh via shell/cmd with argument `--discovery="URL"`
4. Prepare a patched networks.json.
    - First of all, download it from [discovery service](https://discover.spacemesh.io/networks.json)
    - Edit `latestSmappRelease` and set `0.2.2`
    - Edit `smappBaseDownloadUrl` and set `https://storage.googleapis.com/smapp/`
5. Serve it (E.G. `python -m SimpleHTTPServer`), but before
6. Run and try to update :)

Then you can test downgrading the version.

@narayanprusty please, pay attention to the fourth bullet in the list above. It requires some changes in the code that generates or updates the` networks.json` file.

### Artifacts
It is signed to make auto-upgrading works and includes all my previous PRs :)

#### v0.2.1-update.0
- Windows: https://storage.googleapis.com/smapp/v0.2.1-update.0/Spacemesh%20Setup%200.2.1-update.0.exe
- macOS: https://storage.googleapis.com/smapp/v0.2.1-update.0/Spacemesh-0.2.1-update.0.dmg
- Linux: https://storage.googleapis.com/smapp/v0.2.1-update.0/spacemesh_app_0.2.1-update.0_amd64.deb

#### v0.2.2
- Windows: https://storage.googleapis.com/smapp/v0.2.2/Spacemesh%20Setup%200.2.2.exe
- macOS: https://storage.googleapis.com/smapp/v0.2.2/Spacemesh-0.2.2.dmg
- Linux: https://storage.googleapis.com/smapp/v0.2.2/spacemesh_app_0.2.2_amd64.deb

### P.S.
- I've cut the corner. Instead of making snack bar messages, I've added update-related notifications right next to the version, which is visible on any screen. I still like the idea of a snack bar (or even "command center") to put all the messages, errors, and probably even tx logs there, but it is a different big task.
